src/hash_table/chaining.rs

```diff
@@ -0,0 +1,1628 @@
+// Copyright 2019 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Intrusive hash table that uses "separate chaining" for collision resolution.
+#[cfg(any(test, feature = "std"))]
+use std::collections::hash_map::RandomState;
+
+use core::borrow::Borrow;
+use core::cmp;
+use core::hash::{BuildHasher, Hash, Hasher};
+use core::ops::Deref;
+use core::iter;
+use core::fmt;
+
+use crate::adapter::Adapter;
+use crate::array::Array;
+use crate::bits::{Bits, BitsArray};
+use crate::dynamic_array::DynamicArray;
+use crate::hash_adapter::HashAdapter;
+pub use crate::hash_table::ChainedLoadFactor;
+use crate::key_adapter::KeyAdapter;
+use crate::linked_list::Link as LinkedListLink;
+use crate::linked_list::LinkedList;
```

Comment:
You really should be using singly-linked lists here. The memory savings are significant and nobody really cares about iterating through hash tables backwards.

Reply:
I had originally planned for `O(1)` inserts and removals. If I use singly-linked lists, the constant time remove operation is lost.

I had also wanted a general purpose implementation. Currently the max load factor is not constrained; the user is free to create hash tables where the number of buckets is small relative to the number of items. Linear time removals *could* affect this kind of use case.

I have a use case where `link: LinkedListLink` is reused between a double-ended-queue and a hash table to save memory. I'll need an extra pointer if I use singly-linked lists, unless a `LinkedListLink` can be used as a `SinglyLinkedListLink`.

I think the solution is two chaining hash table implementations:

Which names do you like?
- singly-linked list 
  - `SinglyLinkedChainedHashTable`
  - `SinglyChainedHashTable`
  - `SinglyLinkedHashTable`
- linked list
  - `LinkedChainedHashTable`
  - `ChainedHashTable`
  - `LinkedHashTable`

Reply:
`O(1)` removal can be obtained by having cursors point to the previous element in the linked list.

Sharing a link between a linked list and a hash table can be solved by #27 which would allow you to implement your own `Link` types and make them work for different collections.